File: design/swift/HousePour/HPTabs.swift
// House Pour kit — HPTabs, the segmented control (COMPONENTS.md "Controls").
// This is the only segmented control in the look; never a system one.
// `hugging` sizes the control to its content (every segment as wide as the widest label)
// for the floating bottom bar; the default fills its container.

import SwiftUI

public struct HPTabs<Item: Hashable>: View {
    let items: [Item]
    let label: (Item) -> String
    @Binding var selected: Item
    let hugging: Bool
    let trackFill: Color
    let bottomPadded: Bool

    public init(items: [Item], selected: Binding<Item>, hugging: Bool = false,
                trackFill: Color = HPTokens.Colors.bg2, bottomPadded: Bool = true,
                label: @escaping (Item) -> String) {
        self.items = items; _selected = selected; self.hugging = hugging
        self.trackFill = trackFill; self.bottomPadded = bottomPadded; self.label = label
    }

    public var body: some View {
        let track = Capsule(style: .continuous)
        HStack(spacing: HPTokens.Space.tabsGap) {
            ForEach(items, id: \.self) { item in
                let isSelected = item == selected
                Button {
                    withAnimation(HPMotion.color) { selected = item }
                } label: {
                    // The track's vertical inset lives inside the button so the tappable segment spans
                    // the full track height (≥ touchMin); the selected pill stays inset by tabsPad.
                    segmentLabel(item)
                        .padding(.vertical, HPTokens.Space.tabY)
                        .padding(.horizontal, HPTokens.Space.tabX)
                        .frame(maxWidth: hugging ? nil : .infinity)
                        .background(
                            Capsule(style: .continuous)
                                .fill(isSelected ? HPTokens.Colors.panel : Color.clear)
                                .shadow(color: HPTokens.Colors.ink.opacity(isSelected ? HPAlpha.tabShadow : 0),
                                        radius: HPTokens.borderWidth, x: 0, y: HPTokens.borderWidth)
                        )
                        .overlay(
                            Capsule(style: .continuous)
                                .strokeBorder(isSelected ? HPTokens.Colors.line : Color.clear, lineWidth: HPTokens.borderWidth)
                        )
                        .padding(.vertical, HPTokens.Space.tabsPad)
                        .frame(minHeight: HPTokens.Space.touchMin)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityLabel(label(item))
                .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
            }
        }
        .padding(.horizontal, HPTokens.Space.tabsPad)
        .background(track.fill(trackFill))
        .hpBorder(track)
        .padding(.bottom, bottomPadded ? HPTokens.Space.tabsBottom : 0)
    }

    /// In hugging mode every segment sizes to the widest label so widths stay equal
    /// while the control as a whole hugs its content.
    @ViewBuilder private func segmentLabel(_ item: Item) -> some View {
        let isSelected = item == selected
        let text = Text(label(item))
            .hpStyle(HPType.tab, color: isSelected ? HPTokens.Colors.ink : HPTokens.Colors.muted)
            .lineLimit(1)
            .minimumScaleFactor(HPMetric.tabLabelMinScale)
        if hugging {
            ZStack {
                ForEach(items, id: \.self) { other in
                    Text(label(other))
                        .hpStyle(HPType.tab, color: HPTokens.Colors.muted)
                        .lineLimit(1)
                        .minimumScaleFactor(HPMetric.tabLabelMinScale)
                        .hidden()
                }
                text
            }
        } else {
            text
        }
    }
}

public extension HPTabs where Item == String {
    init(items: [String], selected: Binding<String>) {
        self.init(items: items, selected: selected) { $0 }
    }
}

/// The floating bottom tab bar (PRODUCT.md §1): the same segmented control, hugging its
/// content, `panel` fill, hairline line, pill radius, and the one card shadow so it reads
/// as a raised pill over scrolling content. The caller places it `cardGap` above the
/// safe-area bottom and hides it on Sign in, Setup, and inside full-screen viewers.
public struct HPFloatingTabs<Item: Hashable>: View {
    let items: [Item]
    let label: (Item) -> String
    @Binding var selected: Item

    public init(items: [Item], selected: Binding<Item>, label: @escaping (Item) -> String) {
        self.items = items; _selected = selected; self.label = label
    }

    public var body: some View {
        let shape = Capsule(style: .continuous)
        HPTabs(items: items, selected: $selected, hugging: true,
               trackFill: HPTokens.Colors.panel, bottomPadded: false, label: label)
            .hpCardShadow(shape: shape, fill: HPTokens.Colors.panel)
    }
}

File: design/kotlin/housepour/HPTabs.kt
package ca.lucianlabs.housepour

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp

/** The segmented control. `bg2` track, `line` border, pill radius; selected item is a `panel` pill with a contact shadow. */
@Composable
fun HPTabs(
    items: List<String>,
    selected: Int,
    onSelect: (Int) -> Unit,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(HPTokens.Radius.pill)
    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(HPTokens.Colors.bg2, shape)
            .border(HPTokens.BORDER_WIDTH.dp, HPTokens.Colors.line, shape)
            .padding(HPTokens.Space.tabsPad),
        horizontalArrangement = Arrangement.spacedBy(HPTokens.Space.tabsGap),
    ) {
        items.forEachIndexed { index, label ->
            val isSelected = index == selected
            val fill by animateColorAsState(
                targetValue = if (isSelected) HPTokens.Colors.panel else HPTokens.Colors.panel.copy(alpha = 0f),
                animationSpec = tween(HPTokens.Motion.COLOR_MS),
                label = "tabFill",
            )
            val text by animateColorAsState(
                targetValue = if (isSelected) HPTokens.Colors.ink else HPTokens.Colors.muted,
                animationSpec = tween(HPTokens.Motion.COLOR_MS),
                label = "tabText",
            )
            val interaction = remember { MutableInteractionSource() }
            var m = Modifier
                .weight(1f)
            if (isSelected) {
                m = m.hpShadow(HPTokens.Radius.pill, HPShadow(HPTokens.Colors.ink.copy(alpha = 0.12f), 0.dp, 1.dp, 3.dp, 0.dp))
            }
            Box(
                modifier = m
                    .clip(shape)
                    .background(fill, shape)
                    .then(if (isSelected) Modifier.border(HPTokens.BORDER_WIDTH.dp, HPTokens.Colors.line, shape) else Modifier)
                    .selectable(selected = isSelected, interactionSource = interaction, indication = null, role = Role.Tab) { onSelect(index) }
                    .defaultMinSize(minHeight = HPTokens.Space.touchMin)
                    .padding(horizontal = HPTokens.Space.tabX, vertical = HPTokens.Space.tabY),
                contentAlignment = Alignment.Center,
            ) {
                HPText(label, HPTokens.Type.tab, text, maxLines = 1, textAlign = TextAlign.Center)
            }
        }
    }
}
